### PR TITLE
ADR for integration identification

### PR DESCRIPTION
**Pull request by withinfocus**
## 🎟️ Tracking

Originated from bot management discussions amongst leadership.

## 📔 Objective

ADR to enhance our server request pipeline with additional validations and registration of integrations.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: docs/architecture/adr/0023-integration-identifiers.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+Documentation will be provided on the Help Center or this contributing docs site indicating the
+minimally-required request headers for all clients to provide when communicating to the Bitwarden
+platform. Release notes will include mention of this and the future enforcement after three major
+releases. Documentation will also be expanded to provide guidance on how unofficial clients should
+form their provided client version specifically to accurately represent supported "windows" of
+client-server interactivity.
+
+Operations teams will perform the necessary development to validate that required headers are
+present and enable its enforcement after that time; invalid requests will be rejected as a
+`400 Bad Request`.
```
[the remark this answers is not in the document]
Comment (withinfocus) [reply docs/architecture/adr/0023-integration-identifiers.md:78]: 400 is appropriate in my experience, and I see posts like https://stackoverflow.com/a/10576923/553125 suggesting the same. Not sure I understand what those "accommodations" are -- could you elaborate?

[the remark this answers is not in the document]
Comment (withinfocus) [reply docs/architecture/adr/0023-integration-identifiers.md:78]: Who is "we", and where is that done? At the edge? I think it's best we do things the most appropriate way here and expect clients to potentially adjust. I expect our own clients to be fine, and for others to align.

A separate point of missing vs. invalid is interesting to me though, and that could be 400 vs. 403 respectively.

**Review by MGibson1 — COMMENTED** [review]
My biggest concern here is there is nothing in place or specified here that will keep us honest in the future. I **DO NOT** want a system by which we can throttle alternative clients to the point of uselessness.

File: docs/architecture/adr/0023-integration-identifiers.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+form their provided client version specifically to accurately represent supported "windows" of
+client-server interactivity.
+
+Operations teams will perform the necessary development to validate that required headers are
+present and enable its enforcement after that time; requests will be rejected as a `400 Bad Request`
+when required headers are missing and a `403 Forbidden` when provided headers or their values are
+not supported.
+
+A process will be established for integrations to submit support tickets requesting a client
+identifier and secret. Customer Success will work with operations teams to register integrations and
```
Comment (MGibson1) [inline docs/architecture/adr/0023-integration-identifiers.md:82]: I don't see how a secret could be used that is actually kept secret. Is there a clearer plan here?

File: docs/architecture/adr/0023-integration-identifiers.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+status: Proposed
+date: 2024-10-22
+tags: [server]
+---
+
+# 0023 - Identifying Integrated Clients
+
+<AdrTable frontMatter={frontMatter}></AdrTable>
+
+## Context and Problem Statement
```
Comment (MGibson1) [inline docs/architecture/adr/0023-integration-identifiers.md:12]: I don't feel these problems are specific. It's not clear to me how we plan on using these registrations. I think we need to describe what the real benefits of this monitoring and limitations would be.

File: docs/architecture/adr/0023-integration-identifiers.md (hunk 2 above)
Comment (withinfocus) [reply docs/architecture/adr/0023-integration-identifiers.md:82]: Eliminated in a2e7124771e12ec69748c00ed5bcc00883c3d7a6.

File: docs/architecture/adr/0023-integration-identifiers.md [hunk 4]
```diff
@@ -68,22 +68,28 @@ simple addition.
...
 not supported.
 
 A process will be established for integrations to submit support tickets requesting a client
-identifier and secret. Customer Success will work with operations teams to register integrations and
-deliver the needed information. Existing Bitwarden integrations will be issued their own client
-identifiers and secrets. Client details will be provided in requests to the Bitwarden platform.
+identifier. Customer Success will work with operations teams to register integrations and deliver
+the needed information. Existing Bitwarden integrations will be issued their own client identifiers.
+Client details will be provided in requests to the Bitwarden platform, with the goal being the
+identification of scale and conformance per client; attestation and the guarantee that a client is
```
Comment (MGibson1) [inline docs/architecture/adr/0023-integration-identifiers.md:86]:
```suggestion
the needed information. Existing Bitwarden integrations will be issued their own client identifiers.
Client details will be provided in requests to the Bitwarden platform, with the goal being the
identification of conformance per client to the technical needs of the Bitwarden server infrastructure; attestation and the guarantee that a client is
```

**Review by MGibson1 — APPROVED** [review]

**Pull request merged**